File: Engine/ceJavaBinding/lwjglmouse.cc
//
// Created by MCEL on 08.05.2024.
//

#include <ceJavaBinding/lwjglmouse.hh>
#include <ceCore/java_methods.hh>

#define THIS_CLASS "org/crimsonedge/launcher/LwjglMouse"

namespace ce::java
{

LwjglMouse::LwjglMouse()
{
  JNIEnv *pEnv = Java::Get();
  cls = pEnv->FindClass(THIS_CLASS);
  jmethodID constructor = pEnv->GetMethodID(cls, "<init>", "(J)V");
  jobj = pEnv->NewObject(cls, constructor, reinterpret_cast<jlong>(this));
}

void LwjglMouse::SetVisible(bool visible)
{
  static JavaCallVoid1<jboolean> setVisible(Java::Get(), jobj, THIS_CLASS, "setVisible", JAVA_BOOLEAN);
  setVisible.call(Java::Get(), visible);
}

bool LwjglMouse::IsVisible() const
{
  static JavaCallBoolean isVisible(Java::Get(), jobj, THIS_CLASS, "isVisible");
  return isVisible.call(Java::Get(), false);
}

void LwjglMouse::SetCursorMode(eCursorMode mode)
{
  static JavaCallVoid1<jint> setCursorMode(Java::Get(), jobj, THIS_CLASS, "setCursorMode", JAVA_INT);
  setCursorMode.call(Java::Get(), static_cast<std::underlying_type<eCursorMode>::type>(mode));
}

eCursorMode LwjglMouse::GetCursorMode() const
{
  static JavaCallInt getCursorMode(Java::Get(), jobj, THIS_CLASS, "getCursorMode");
  switch (getCursorMode.call(Java::Get(), 0))
  {
    case 0:
      return eCursorMode::Free;
    case 1:
      return eCursorMode::Confined;
    case 2:
      return eCursorMode::Fixed;
  }
  return eCursorMode::Free;
}


int32_t LwjglMouse::GetX() const
{
  static JavaCallInt getX(Java::Get(), jobj, THIS_CLASS, "getX");
  return getX.call(Java::Get(), 0);
}

int32_t LwjglMouse::GetY() const
{
  static JavaCallInt getY(Java::Get(), jobj, THIS_CLASS, "getY");
  return getY.call(Java::Get(), 0);
}

int32_t LwjglMouse::GetDeltaX() const
{
  static JavaCallInt getDeltaX(Java::Get(), jobj, THIS_CLASS, "getDeltaX");
  return getDeltaX.call(Java::Get(), 0);
}

int32_t LwjglMouse::GetDeltaY() const
{
  static JavaCallInt getDeltaY(Java::Get(), jobj, THIS_CLASS, "getDeltaY");
  return getDeltaY.call(Java::Get(), 0);
}

int32_t LwjglMouse::GetWheel() const
{
  static JavaCallInt getWheel(Java::Get(), jobj, THIS_CLASS, "getWheel");
  return getWheel.call(Java::Get(), 0);
}
int32_t LwjglMouse::GetWheelHorizontal() const
{
  static JavaCallInt getWheelHorizontal(Java::Get(), jobj, THIS_CLASS, "getWheelHorizontal");
  return getWheelHorizontal.call(Java::Get(), 0);
}

bool LwjglMouse::IsButtonDown(eMouseButton mouseButton) const
{
  static JavaCallBoolean1<jint> isButtonDown(Java::Get(), jobj, THIS_CLASS, "isButtonDown", JAVA_INT);
  return isButtonDown.call(Java::Get(), static_cast<std::underlying_type<eMouseButton>::type>(mouseButton), false);
}

bool LwjglMouse::IsButtonUp(eMouseButton mouseButton) const
{
  static JavaCallBoolean1<jint> isButtonUp(Java::Get(), jobj, THIS_CLASS, "isButtonUp", JAVA_INT);
  return isButtonUp.call(Java::Get(), static_cast<std::underlying_type<eMouseButton>::type>(mouseButton), true);
}
bool LwjglMouse::IsButtonPressed(eMouseButton mouseButton) const
{
  static JavaCallBoolean1<jint> isButtonPressed(Java::Get(), jobj, THIS_CLASS, "isButtonPressed", JAVA_INT);
  return isButtonPressed.call(Java::Get(), static_cast<std::underlying_type<eMouseButton>::type>(mouseButton), false);
}
bool LwjglMouse::IsButtonReleased(eMouseButton mouseButton) const
{
  static JavaCallBoolean1<jint> isButtonReleased(Java::Get(), jobj, THIS_CLASS, "isButtonReleased", JAVA_INT);
  return isButtonReleased.call(Java::Get(), static_cast<std::underlying_type<eMouseButton>::type>(mouseButton), false);
}

}

File: Java/JCrimsonEdge/CeLWJGLWindow/src/main/java/org/crimsonedge/lwjgl/LwjglMouse.java
package org.crimsonedge.lwjgl;

import org.crimsonedge.core.input.ECursorMode;
import org.crimsonedge.core.input.EMouseButton;
import org.crimsonedge.core.input.IMouse;

import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;

public class LwjglMouse implements IMouse {

    private final long ref;

    private int posX;

    private int posY;

    private int lastX;
    private int lastY;

    private boolean visible = true;
    private ECursorMode cursorMode = ECursorMode.FREE;
    private Robot robot;

    public static final class Event {
        int button;
        boolean down;
    }

    private Event[] events = new Event[1000];
    private int numEvents = 0;

    private boolean[] buttons = new boolean[EMouseButton.values().length];
    private boolean[] prevButtons = new boolean[EMouseButton.values().length];

    private LwjglWindowCanvas canvas;
    private Cursor blankCursor;

    public LwjglMouse(long ref) {
        this.ref = ref;
        PointerInfo inf = MouseInfo.getPointerInfo();
        Point p = inf.getLocation();
        posX = lastX = p.x;
        posY = lastY = p.y;
        for (int i = 0; i < events.length; i++) {
            events[i] = new Event();
        }
        try {
            robot = new Robot();
        }
        catch (Exception e) {
            e.printStackTrace();
        }

        BufferedImage cursorImg = new BufferedImage(16, 16, BufferedImage.TYPE_INT_ARGB);

        this.blankCursor = Toolkit.getDefaultToolkit().createCustomCursor(cursorImg, new Point(0, 0), "blank cursor");
    }

    public LwjglWindowCanvas getCanvas() {
        return canvas;
    }

    public void setCanvas(LwjglWindowCanvas canvas) {
        this.canvas = canvas;
    }

    @Override
    public void setVisible(boolean visible) {
        this.visible = visible;
        if (visible) {
            this.canvas.setCursor(new Cursor(Cursor.DEFAULT_CURSOR));
        }
        else {
            this.canvas.setCursor(this.blankCursor);
        }
    }

    @Override
    public boolean isVisible() {
        return this.visible;
    }

    public void setCursorMode(int mode) {
        setCursorMode(ECursorMode.values()[mode]);
    }

    @Override
    public void setCursorMode(ECursorMode mode) {
        if (this.cursorMode != mode) {
            this.cursorMode = mode;
            if (this.cursorMode == ECursorMode.FIXED) {
                Point center = new Point(this.canvas.getWidth() / 2, this.canvas.getHeight() / 2);
                SwingUtilities.convertPointToScreen(center, this.canvas);

                this.lastX = center.x;
                this.lastY = center.y;
                this.robot.mouseMove(lastX, lastY);
                this.posX = this.lastX;
                this.posY = this.lastY;
            }
        }
    }

    @Override
    public ECursorMode getCursorMode() {
        return ECursorMode.FREE;
    }

    @Override
    public int getX() {
        return this.posX;
    }

    @Override
    public int getY() {
        return this.posY;
    }

    @Override
    public int getDeltaX() {
        return this.posX - lastX;
    }

    @Override
    public int getDeltaY() {
        return this.posY - lastY;
    }

    @Override
    public int getWheel() {
        return 0;
    }

    @Override
    public int getWheelHorizontal() {
        return 0;
    }

    public boolean isButtonDown(int mouseButton) {
        return buttons[mouseButton];
    }

    public boolean isButtonUp(int mouseButton) {
        return !buttons[mouseButton];
    }

    public boolean isButtonPressed(int mouseButton) {
        return buttons[mouseButton] && !prevButtons[mouseButton];
    }

    public boolean isButtonReleased(int mouseButton) {
        return !buttons[mouseButton] && prevButtons[mouseButton];
    }

    @Override
    public boolean isButtonDown(EMouseButton mouseButton) {
        return isButtonDown(mouseButton.ordinal());
    }

    @Override
    public boolean isButtonUp(EMouseButton mouseButton) {
        return isButtonUp(mouseButton.ordinal());
    }

    @Override
    public boolean isButtonPressed(EMouseButton mouseButton) {
        return isButtonPressed(mouseButton.ordinal());
    }

    @Override
    public boolean isButtonReleased(EMouseButton mouseButton) {
        return isButtonReleased(mouseButton.ordinal());
    }

    public void pushEvent(int button, boolean down) {
        Event event = events[numEvents++];
        event.button = button;
        event.down = down;
    }

    public void update() {
        updateButtons();
        updatePosition();
    }

    private void updateButtons() {
        System.arraycopy(this.buttons, 0, this.prevButtons, 0, this.buttons.length);
        for (int i = 0; i < this.numEvents; i++) {
            Event evt = events[i];
            this.buttons[evt.button] = evt.down;
        }
        this.numEvents = 0;
    }

    private void updatePosition() {
        PointerInfo inf = MouseInfo.getPointerInfo();
        Point p = inf.getLocation();

        if (this.cursorMode == ECursorMode.FIXED) {
            if (this.robot != null) {
                posX = p.x;
                posY = p.y;

                Point center = new Point(this.canvas.getWidth() / 2, this.canvas.getHeight() / 2);
                SwingUtilities.convertPointToScreen(center, this.canvas);

                this.lastX = center.x;
                this.lastY = center.y;
                this.robot.mouseMove(lastX, lastY);
                return;
            }
        }
        else if (this.cursorMode == ECursorMode.CONFINED) {
            if (this.robot != null) {
                Point topLeft = new Point(0, 0);
                Point bottomRight = new Point(this.canvas.getWidth(), this.canvas.getHeight());
                SwingUtilities.convertPointToScreen(topLeft, this.canvas);
                SwingUtilities.convertPointToScreen(bottomRight, this.canvas);

                int newX = Math.min(Math.max(p.x, topLeft.x), bottomRight.x);
                int newY = Math.min(Math.max(p.y, topLeft.y), bottomRight.y);
                if (newX != p.x || newY != p.y) {
                    this.robot.mouseMove(newX, newY);
                }

                lastX = posX;
                lastY = posY;
                posX = newX;
                posY = newY;
                return;
            }
        }

        lastX = posX;
        lastY = posY;
        posX = p.x;
        posY = p.y;
    }


}
